superset/common/query_object.py

```diff
@@ -396,6 +398,24 @@ def cache_key(self, **extra: Any) -> str:
         if annotation_layers:
             cache_dict["annotation_layers"] = annotation_layers
 
+        # Add an impersonation key to cache if impersonation is enabled on the db
+        if (
+            feature_flag_manager.is_feature_enabled("CACHE_IMPERSONATION")
+            and self.datasource
+            and hasattr(self.datasource, "database")
+            and self.datasource.database.impersonate_user
+        ):
+
+            if key := self.datasource.database.db_engine_spec.get_impersonation_key(
+                get_username()
```

Comment:
Passing the user object here instead of the username makes the method more flexible, since other DBs might want to use the ID or the email. You can simply do:

```suggestion
                getattr(g, "user", None)
```

And add `from flask import g` to the top of the file.

Reply:
Thanks @betodealmeida for the review, I don't know how I had missed that `g.user` is an instance of `User` model 😅 .

I made the change, please take a look again.